src/repr/src/adt/timestamp.rs

```diff
@@ -0,0 +1,579 @@
+            return Err(TimestampError::OutOfRange);
+        }
+        Ok(Self { t })
+    }
+
+    pub fn checked_add_signed(self, rhs: Duration) -> Option<T> {
+        self.t.checked_add_signed(rhs)
+    }
+
+    pub fn checked_sub_signed(self, rhs: Duration) -> Option<T> {
+        self.t.checked_sub_signed(rhs)
+    }
+}
+
+impl TryFrom<NaiveDateTime> for CheckedTimestamp<NaiveDateTime> {
+    type Error = TimestampError;
+
+    fn try_from(value: NaiveDateTime) -> Result<Self, Self::Error> {
+        Self::from_timestamplike(value)
+    }
+}
+
+impl TryFrom<DateTime<Utc>> for CheckedTimestamp<DateTime<Utc>> {
+    type Error = TimestampError;
+
+    fn try_from(value: DateTime<Utc>) -> Result<Self, Self::Error> {
+        Self::from_timestamplike(value)
+    }
+}
+
+// TODO: Should we just use .into instead of this?
+impl<T: TimestampLike> std::ops::Deref for CheckedTimestamp<T> {
+    type Target = T;
+
+    #[inline]
+    fn deref(&self) -> &T {
+        &self.t
+    }
+}
+
+impl From<CheckedTimestamp<NaiveDateTime>> for NaiveDateTime {
+    fn from(val: CheckedTimestamp<NaiveDateTime>) -> Self {
+        val.t
+    }
+}
+
+impl From<CheckedTimestamp<DateTime<Utc>>> for DateTime<Utc> {
+    fn from(val: CheckedTimestamp<DateTime<Utc>>) -> Self {
+        val.t
+    }
+}
+
+impl CheckedTimestamp<NaiveDateTime> {
+    pub fn to_naive(&self) -> NaiveDateTime {
+        self.t
+    }
+}
+
+impl CheckedTimestamp<DateTime<Utc>> {
+    pub fn to_naive(&self) -> NaiveDateTime {
+        self.t.date().naive_utc().and_time(self.t.time())
+    }
+}
+
+impl Display for CheckedTimestamp<NaiveDateTime> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.t.fmt(f)
+    }
+}
+
+impl Display for CheckedTimestamp<DateTime<Utc>> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.t.fmt(f)
+    }
+}
+
+impl RustType<ProtoNaiveDateTime> for CheckedTimestamp<NaiveDateTime> {
+    fn into_proto(&self) -> ProtoNaiveDateTime {
+        self.t.into_proto()
+    }
+
+    fn from_proto(proto: ProtoNaiveDateTime) -> Result<Self, TryFromProtoError> {
+        Ok(Self {
+            t: NaiveDateTime::from_proto(proto)?,
+        })
+    }
+}
+
+impl RustType<ProtoNaiveDateTime> for CheckedTimestamp<DateTime<Utc>> {
+    fn into_proto(&self) -> ProtoNaiveDateTime {
+        self.t.into_proto()
+    }
+
+    fn from_proto(proto: ProtoNaiveDateTime) -> Result<Self, TryFromProtoError> {
+        Ok(Self {
+            t: DateTime::<Utc>::from_proto(proto)?,
+        })
+    }
+}
+
+impl<T: Sub<Output = Duration>> Sub<CheckedTimestamp<T>> for CheckedTimestamp<T> {
+    type Output = Duration;
+
+    #[inline]
+    fn sub(self, rhs: CheckedTimestamp<T>) -> Duration {
+        self.t - rhs.t
+    }
+}
+
+impl<T: Sub<Duration, Output = T>> Sub<Duration> for CheckedTimestamp<T> {
+    type Output = T;
+
+    #[inline]
+    fn sub(self, rhs: Duration) -> T {
+        self.t - rhs
+    }
+}
```

Comment:
this file all seems reasonable to me...the extension traits are a good pattern, and it seems the general process is that they require the times and dates to act like a `chrono::NaiveDate`, or at least be able to generate one